src/axom/mir/detail/elvira_detail.hpp

```diff
@@ -0,0 +1,382 @@
+ * \param[out] range Start and end points of the range that can be used to
+ *                   make a clipping plane for the shape.
+ */
+template <typename ShapeType, typename T, int NDIMS>
+AXOM_HOST_DEVICE inline void computeRange(const ShapeType &shape,
+                                          const axom::primal::Vector<T, NDIMS> &normal,
+                                          axom::primal::Point<T, NDIMS> range[2])
+{
+  // Compute the shape bounding box.
+  const auto bbox = axom::primal::compute_bounding_box(shape);
+
+  const axom::primal::Plane<T, NDIMS> P(normal, bbox.getCentroid(), false);
+
+  // Compute distances from all points in shape to plane.
+  range[0] = range[1] = bbox.getCentroid();
+  double dist[2] = {0., 0.};
+  for(axom::IndexType ip = 0; ip < shape.numVertices(); ip++)
+  {
+    const auto d = static_cast<T>(P.signedDistance(shape[ip]));
+    if(d < dist[0])
+    {
+      dist[0] = d;
+      range[0] = bbox.getCentroid() + (normal * d);
+    }
+    if(d > dist[1])
+    {
+      dist[1] = d;
+      range[1] = bbox.getCentroid() + (normal * d);
+    }
+  }
+}
+
+/*!
+ * \brief Clip a shape until a specified volume is achieved, up to a tolerance
+ *        or max number of iterations. The clipped shape is returned. The
+ *        origin of the clipping plane is returned as an out argument.
+ *
+ * \param shape The shape to be clipped.
+ * \param normal The normal along which the clipping plane will move.
+ * \param _range The start, end points of a range that contains the plane origin.
+ * \param matVolume The target shape volume.
+ * \param max_iterations The max number of iterations to get the answer.
+ * \param tolerance The allowable volume tolerance.
+ * \param[out] pt The origin of the clipping plane that was used.
+ */
+template <typename ClipResultType, typename ShapeType, typename T, int NDIMS>
+AXOM_HOST_DEVICE inline ClipResultType clipToVolume(
+  const ShapeType &shape,
+  const axom::primal::Vector<T, NDIMS> &normal,
+  const axom::primal::Point<T, NDIMS> _range[2],
+  double matVolume,
+  int max_iterations,
+  double tolerance,
+  axom::primal::Point<T, NDIMS> &pt)
+{
+  namespace bputils = axom::mir::utilities::blueprint;
+  // The range for the interval
+  axom::primal::Point<T, NDIMS> range[2] = {_range[0], _range[1]};
+  // This array holds the volumes for the interval.
+  double f_t[2] = {0., matVolume};
+  // The blend value within the current interval.
+  double t_blend = 0.5;
+
+  ClipResultType clippedShape {};
+  for(int iterations = 0; iterations < max_iterations; iterations++)
+  {
+    // Pick the middle of the range and position the plane there.
+    pt = axom::primal::Point<T, NDIMS>::lerp(range[0], range[1], t_blend);
+
+    // The ELVIRA normals point away from the material. Axom's clipping
+    // keeps the shape where the normal points into the shape. Reverse
+    // the ELVIRA normal when forming the plane so we keep the right piece.
```

Comment:
👍

Reply:
Minor: should the normal be flipped outside the loop?

Reply:
Done.